File: Longest_Substring_Without_Repeating_Characters.go
package main

import "fmt"

func lengthOfLongestSubstring(s string) int {
	cMap := make(map[byte]bool)
	res := 0
	for i, j := 0, 0; j < len(s); {
		if !cMap[s[j]] {
			cMap[s[j]] = true
			j++
			res = max(res, j-i)
		} else {
			for s[i] != s[j] {
				cMap[s[i]] = false
				i++
			}
			cMap[s[i]] = false
			i++
		}
	}
	return res
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func main() {
	testCases := []string{"abcabcbb", "bbbbb", "pwwkew"}
	for _, testCase := range testCases {
		fmt.Println(lengthOfLongestSubstring(testCase))
	}
}

File: Longest_Substring_Without_Repeating_Characters.cpp
#include <iostream>
#include <set>
#include <string>
#include <vector>
using namespace std;
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int left = 0, right = 0, maxLength = 0, curLength = 0;
        set<char> charSet;
        while (right < s.length()) {
            while (charSet.find(s[right]) != charSet.end()) {
                charSet.erase(s[left]);
                left++;
            }

            charSet.insert(s[right]);
            curLength = right - left + 1;
            maxLength = max(maxLength, curLength);
            right++;
            // cout << left << " " << right << " " << curLength << endl;
        }
        return maxLength;
    }
};

int main() {
    vector<string> testCases = {"abcabcbb", "bbbb", "pwwkew", "tmmzuxt"};
    Solution sol;
    for (auto testCase : testCases) {
        cout << sol.lengthOfLongestSubstring(testCase) << endl;
    }
    return 0;
}